content/ember/v3/array-observers.md

```diff
@@ -0,0 +1,345 @@
+rearrange its own representation of the array. A simplified version of this
+logic looks like the following:
+
+```js
+export default Component.extend({
+  layout: layout,
+
+  init() {
+    this._cells = A();
+  },
+
+  _needsRevalidate(){
+    if (this.isDestroyed || this.isDestroying) {return;}
+    this.rerender();
+  },
+
+  didReceiveAttrs() {
+    this._super();
+
+    this.updateItems();
+  },
+
+  updateItems(){
+    var rawItems = this.get('items');
+
+    if (this._rawItems !== rawItems) {
+      if (this._items && this._items.removeArrayObserver) {
+        this._items.removeArrayObserver(this, {
+          willChange: noop,
+          didChange: '_needsRevalidate'
+        });
+      }
+      this._rawItems = rawItems;
+      var items = A(rawItems);
+      this.set('_items', items);
+
+      if (items && items.addArrayObserver) {
+        items.addArrayObserver(this, {
+          willChange: noop,
+          didChange: '_needsRevalidate'
+        });
+      }
+    }
+  },
+
+  willRender() {
+    this.updateCells();
+  },
+
+  updateCells() {
+    // ...
+  },
+
+  actions: {
+    scrollChange(scrollLeft, scrollTop) {
+      // ...
+      if (scrollLeft !== this._scrollLeft ||
+          scrollTop !== this._scrollTop) {
+        set(this, '_scrollLeft', scrollLeft);
+        set(this, '_scrollTop', scrollTop);
+        this._needsRevalidate();
+      }
+    },
+    clientSizeChange(clientWidth, clientHeight) {
+      if (this._clientWidth !== clientWidth ||
+          this._clientHeight !== clientHeight) {
+        set(this, '_clientWidth', clientWidth);
+        set(this, '_clientHeight', clientHeight);
+        this._needsRevalidate();
+      }
+    }
+  }
+});
+```
+
+We can refactor this to update the cells themselves when they are accessed, by
+accessing them into a computed property that depends on the items array, and
+which updates the cells when it is accessed:
+
+```js
+export default Component.extend({
+  layout: layout,
+
+  init() {
+    this._cells = A();
+  },
+
+  cells: computed('items.[]', function() {
+    this.updateCells();
+
+    return this._cells;
+  })
+
+  updateCells() {
+    // ...
+  },
+
+  actions: {
+    scrollChange(scrollLeft, scrollTop) {
+      // ...
+      if (scrollLeft !== this._scrollLeft ||
+          scrollTop !== this._scrollTop) {
+        set(this, '_scrollLeft', scrollLeft);
+        set(this, '_scrollTop', scrollTop);
+        this.notifyPropertyChange('cells');
+      }
+    },
+    clientSizeChange(clientWidth, clientHeight) {
+      if (this._clientWidth !== clientWidth ||
+          this._clientHeight !== clientHeight) {
+        set(this, '_clientWidth', clientWidth);
+        set(this, '_clientHeight', clientHeight);
+        this.notifyPropertyChange('cells');
+      }
+    }
+  }
+});
+```
+
+Mutating untracked local state like this is generally ok as long as the state is
+essentially a cached representation of computed or getter is deriving in
```

Comment:
I was confused by "getter is deriving". Was "is" meant to be included in the sentence?